### ISSUE TITLE
Uninitialized self._dependents in unsubscribe

### ISSUE THREAD

**Issue by jensens** [zopefoundation/zope.interface#165]
`plone.supermodel` uses interfaces very dynamic. So, here it constructs schemas/interfaces and so it comes to an unsubscribe when the 
```
Error in test test_dynamic_schema_refreshed_on_modify_schema_policy (plone.dexterity.tests.test_fti.TestFTIEvents)
Traceback (most recent call last):
  File "/home/jensens/.pyenv/versions/3.7.5/lib/python3.7/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/home/jensens/.pyenv/versions/3.7.5/lib/python3.7/unittest/case.py", line 628, in run
    testMethod()
  File "/home/jensens/ws/cdev/p52/src/plone.dexterity/plone/dexterity/tests/test_fti.py", line 749, in test_dynamic_schema_refreshed_on_modify_schema_policy
    DexterityFTIModificationDescription('schema_policy', '')
  File "/home/jensens/ws/cdev/p52/src/plone.dexterity/plone/dexterity/fti.py", line 561, in ftiModified
    syncSchema(model.schema, schema, overwrite=True, sync_bases=sync_bases)
  File "/home/jensens/ws/cdev/p52/src/plone.supermodel/plone/supermodel/utils.py", line 342, in syncSchema
    dest.__bases__ = tuple(bases)
  File "/home/jensens/ws/cdev/p52/src/zope.interface/src/zope/interface/interface.py", line 255, in __setBases
    b.unsubscribe(self)
  File "/home/jensens/ws/cdev/p52/src/zope.interface/src/zope/interface/interface.py", line 242, in unsubscribe
    n = self._dependents[dependent]
  File "/home/jensens/.pyenv/versions/3.7.5/lib/python3.7/weakref.py", line 396, in __getitem__
    return self.data[ref(key)]
KeyError: <weakref at 0x7fea75b7a5f0; to 'InterfaceClass' at 0x7fea75b40830 (IBlank)>
```

I propose to initialise the `_dependents` in unsubscribe first before, if this was not done before.

```
@@ -235,6 +238,8 @@ class Specification(SpecificationBase):
         self._dependents[dependent] = self.dependents.get(dependent, 0) + 1
 
     def unsubscribe(self, dependent):
+        if self._dependents is None:
+            self._dependents = weakref.WeakKeyDictionary()
         try:
             n = self._dependents[dependent]
         except TypeError:
```

I'll prepare a PR.

**Comment by jamadden** [comment]
So it's unsubscribing something that never was subscribed. That doesn't seem right….

> I propose to initialise the _dependents in unsubscribe first before, if this was not done before.

I'd suggest, instead, simply making that a no-op. 

```python
def unsubscribe(self, dependent):
   if not self._dependents:
        # The caller is attempting the impossible: 
        # unsubscribing something never subscribed.
        return
    try:
        # as existing
    …
```

**Comment by jamadden** [comment]
Ow, nope, sorry, not a no-op, it should actually raise `KeyError`.  I'd still prefer doing that manually to initialising the data and taking the memory that may never be used.

**Comment by jamadden** [comment]
Wait, no, I'm still not getting it. `None[key]` raises the `TypeError` which is turned into the `KeyError`. zope.interface 4.7.1 would have raised a `KeyError` in the same circumstance. Externally, the visible behaviour should be exactly the same now as in 4.7.1.

**Comment by jensens** [comment]
But it isn't. Maybe this is the wrong place. Then it is upon __setBases to check if an unsubscribe is possible?

**Comment by jensens** [comment]
like so?

``` python
@@ -251,7 +249,10 @@ class Specification(SpecificationBase):
     def __setBases(self, bases):
         # Remove ourselves as a dependent of our old bases
         for b in self.__bases__:
-            b.unsubscribe(self)
+            try:
+                b.unsubscribe(self)
+            except KeyError:
+                pass
 
         # Register ourselves as a dependent of our bases
```

**Comment by jamadden** [comment]
> Maybe this is the wrong place. Then it is upon __setBases to check if an unsubscribe is possible?

Ahh. I was just reading the plone.dexterity/supermodel test case. `__setBases` makes more sense as the source of the issue. But I suspect there's a bug lurking in plone.dexterity/supermodel where it's not correctly initializing a super class because if it had properly set `__bases__` before they should have been subscribed…Or perhaps it's overriding the `__bases__` property to return things that it never actually told the super class about?

I would suggest that the fix would be in one of those places, making them properly inform the superclass about the bases or properly subscribing them. (If they're not subscribed caches will get out of sync).

Alternately, we could just iterate across `self._bases` instead of invoking the property, but that doesn't seem right (for the same reason that it's possibly not right to directly check `if self._dependents:` in this function).

It's not clear to me that adding a `try/except: pass` here is correct. That also seems like a recipe for bugs and things getting out of sync. It's like disabling foreign key constraints in SQL: not a good idea. However, it's during *removal* not addition, and if we were never subscribed in the first place, I don't know what could practically go wrong if we ignore the issue. Whatever damage was going to happen probably already did.

So to sum up, my thoughts would be:

1. this feels like maybe some other place is breaking the rules and hopefully that's easy to find and fix;
2. If I'm completely off base there, or it turns out to be super impossible to fix the source of the problem, it's probably not completely bad to ignore the exception here. But I would hope for a comment about exactly why we have to do that, if we did find some project breaking the rules

**Comment by jamadden** [comment]
I was giving some thought to how a subclass or dynamic implementation might even get into this state, and I'm having trouble coming up with a scenario. The closest I've got so far was creating a "clone" of an object using, say, `clone = type(Class.__providedBy__).__new__(type(Class.__providedBy__))` and then attempting to copy state using `clone.__dict__ = dict(Class.__providedBy__.__dict__)`. That just won't work because most of the state isn't in the `__dict__` anymore, it's in `__slots__` (or the C struct). If you override `__bases__` to return something else and not touch `_bases`…well, you blow up with an AttributeError.

**Comment by jensens** [comment]
I debugged into the test by catching the keyerror:

```
[10] > /home/jensens/ws/cdev/p52/src/zope.interface/src/zope/interface/interface.py(259)__setBases()
-> raise
   2 frames hidden (try 'help hidden_frames')
(Pdb++) ll
 252         def __setBases(self, bases):                                                                                                                                                                          
 253             # Remove ourselves as a dependent of our old bases                                                                                                                                                
 254             for b in self.__bases__:                                                                                                                                                                          
 255                 try:                                                                                                                                                                                          
 256                     b.unsubscribe(self)                                                                                                                                                                       
 257                 except KeyError:                                                                                                                                                                              
 258                     import pdb; pdb.set_trace()                                                                                                                                                               
 259  ->                 raise                                                                                                                                                                                     
 260                                                                                                                                                                                                               
 261             # Register ourselves as a dependent of our bases                                                                                                                                                  
 262             self._bases = bases                                                                                                                                                                               
 263             for b in bases:                                                                                                                                                                                   
 264                 b.subscribe(self)                                                                                                                                                                             
 265             self.changed(self)                                                                                                                                                                                (Pdb++) b
<InterfaceClass zope.interface.Interface>
(Pdb++) self
<InterfaceClass plone.dexterity.tests.test_fti.IBlank>
(Pdb++) IBlank in Interface._dependents
False

```

I read this, the `Interface` base class (or class-like-instance) has not set `IBlank` as an `_dependent`. But it should because `IBlank` inherits from `Interface`:

https://github.com/plone/plone.dexterity/blob/a8a234d3242f2abbf30f22cea321a3eaa299f5f8/plone/dexterity/tests/test_fti.py#L714-L715

Ok, lets dig deeper.

I put a pdb right after the definition of `IBlank`:
```
(Pdb++) ll
 707         def test_dynamic_schema_refreshed_on_modify_schema_policy(self):                                                                                                                                      
 708             portal_type = u"testtype"                                                                                                                                                                         
 709             fti = DexterityFTI(portal_type)                                                                                                                                                                   
 710                                                                                                                                                                                                               
 711             class INew(Interface):                                                                                                                                                                            
 712                 title = zope.schema.TextLine(title=u"title")                                                                                                                                                  
 713                                                                                                                                                                                                               
 714             class IBlank(Interface):                                                                                                                                                                          
 715                 pass                                                                                                                                                                                          
 716                                                                                                                                                                                                               
 717             import pdb; pdb.set_trace()                                                                                                                                                                       
 718  ->         class TestSchemaPolicy(DexteritySchemaPolicy):                                                                                                                                                    
 719                 def bases(self, schemaName, tree):                                                                                                                                                            
 720                     return (INew,)                                                                                                                                                                            
 721                                                                                                                                                                                                               
 722             gsm = getGlobalSiteManager()            
(Pdb++) IBlank in Interface._dependents
False                                                                                                                                                         
```

Wow, what is going on here? I double checked if `Interface` was overwritten somewhere, but I can not find anything. Also Interface looks fine:

```
(Pdb++) len(Interface._dependents)
476
(Pdb++) Interface
<InterfaceClass zope.interface.Interface>
(Pdb++) Interface.__class__
<class 'zope.interface.interface.InterfaceClass'>
(Pdb++) from zope.interface import Interface as original
(Pdb++) original is Interface
True
```

Lets try to define a subclass and lets see if it registers as `_dependent`:

```
(Pdb++) class IFoo(Interface): pass
(Pdb++) IFoo in Interface._dependents
True
```

I am a bit out of ideas what could go wrong here....

**Comment by jensens** [comment]
Ok, I am one step further: The `IBlank` is defined as a local class in 4 tests, all of them get the `` __module__` "'plone.dexterity.tests.test_fti" and the name "IBlank". Thus all of them have the same hash, but in fact they are different.

So I suppose this is a bug in the module name detection here in line 367:

https://github.com/zopefoundation/zope.interface/blob/be8275645900d3d598d59ef9644983a77ccc8f70/src/zope/interface/interface.py#L352-L369

It anyway looks a bit like a piece of black magic.

**Comment by jensens** [comment]
A quick check if using ``__qualname__`` from attrs for the ``__hash__`` instead of ``__name__`` was successful.

```
diff --git src/zope/interface/interface.py src/zope/interface/interface.py
index 57998e0..6b136b9 100644
--- src/zope/interface/interface.py
+++ src/zope/interface/interface.py
@@ -370,6 +370,10 @@ class InterfaceClass(Element, InterfaceBase, Specification):
 
         self.__module__ = __module__
 
+        # local class definitions may share the same module and name,
+        # use qualname for hashes
+        self.__qualname__ = attrs.get('__qualname__', name)
+
         d = attrs.get('__doc__')
         if d is not None:
             if not isinstance(d, Attribute):
@@ -591,7 +595,7 @@ class InterfaceClass(Element, InterfaceBase, Specification):
         try:
             return self._v_cached_hash
         except AttributeError:
-            self._v_cached_hash = hash((self.__name__, self.__module__))
+            self._v_cached_hash = hash((self.__qualname__, self.__module__))
         return self._v_cached_hash
 
     def __eq__(self, other):
```

Question now is, should this reflect at other places too? 

- `__identifier__`
- `__repr__`
- ...?

I think this would break lots of tests in lots of packages and many people would not consider us friends any more if we break this ;-) ....

Also, as far as I now it needs Python 3.3+.

**Comment by mgedmin** [comment]
Uh, I'm sure I'm not following the details exactly, but if zope.interface has trouble with different dynamically constructed interfaces sharing the same name, that might be a problem for my code.

(What I'm doing:

```python
def parse_dropdowns(csv_text, bases=(), _cache={}):
    ...
    schema = InterfaceClass('dynamic-dropdowns', bases, field_dict)
    return schema
```
plus there's some caching so the same csv_text+bases pair always returns the same interface object. 
 These schemas are then used to generate forms.)

Note that `__qualname__ ` would still be the same for different interface objects in my code example.

**Comment by mgedmin** [comment]
> Ok, I am one step further: The IBlank is defined as a local class in 4 tests, all of them get the __module__ "'plone.dexterity.tests.test_fti" and the name "IBlank". Thus all of them have the same hash, but in fact they are different.

It should not be a problem: the way hashing works, different objects are allowed to have the same hash (with some potential efficiency loss).  The only requirement is for equal objects to have the same hash.

**Comment by jensens** [comment]
I added failing test now in branch https://github.com/zopefoundation/zope.interface/tree/issue165-test where the tests are failing, see https://travis-ci.org/zopefoundation/zope.interface/builds/646349843

> It should not be a problem: the way hashing works, different objects are allowed to have the same hash (with some potential efficiency loss). The only requirement is for equal objects to have the same hash.

But if you use those objects with the same hash as a key in dict, the second object is considered to be already in used in the dict - this is what happens in `Interface._dependents`. Totally different objects are registered as dependents, but only one lands in the WeakKeyDictionary.

**Comment by mgedmin** [comment]
> But if you use those objects with the same hash as a key in dict, the second object is considered to be already used in the dict - this is what happens in Interface._dependents. Totally different objects are registered as dependents, but only one lands in the WeakKeyDictionary.

That is because of `__eq__`, not because of `__hash__`.

Incidentally, I'm now looking at zope.interface.InterfaceClass and I see this comment at the top of `__cmp`:

```
        # Yes, I did mean to name this __cmp, rather than __cmp__.
        # It is a private method used by __lt__ and __gt__.
        # I don't want to override __eq__ because I want the default
        # __eq__, which is really fast.
```

but then `__eq__` is overridden anyway (to rely on `__cmp`).  The comment is from 2007, the `__eq__` method is from 2011.  Welp.

**Comment by jensens** [comment]
As far as I know Python, the default `__eq__` would compare the hashes based on identity. Also the default `__hash__` is a hash of `id(self)` (or the uses the id direct - I am not sure, but it does not matter much). 

The idea of having a `__hash__` different from `id` is probably to have be able to persist/pickle the object and have the same id afterwards. 

So, when `__hash__` was overriden, then `__eq__` needs to follow the same logic.

So, IMO the comment is just outdated and should be updated.

Also, our equality check could be faster by comparing the own and other `__hash__`.

**Comment by jamadden** [comment]
> Also, our equality check could be faster by comparing the own and other __hash__.

Equal hash codes do not imply equal objects. You'd still have to do the comparison so it would slow down the equal case.
```python
def __eq__(self, other):
    if hash(self) == hash(other):
        return self.__cmp(other) == 0
    return False
```

**Comment by jensens** [comment]
invalid
